crates/oxc_linter/src/rules/typescript/no_dynamic_delete.rs

```diff
@@ -0,0 +1,162 @@
+
+#[derive(Debug, Default, Clone)]
+pub struct NoDynamicDelete;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow using the delete operator on computed key expressions.
+    ///
+    /// ### Why is this bad?
+    /// Deleting dynamically computed keys can be dangerous and in some cases not well optimized.
+    /// Using the delete operator on keys that aren't runtime constants could be a sign that you're using the wrong data structures.
+    /// Consider using a Map or Set if you’re using an object as a key-value collection.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// const container: { [i: string]: 0 } = {};
+    /// delete container['aa' + 'b'];
+    /// ```
+    NoDynamicDelete,
+    restriction,
+);
+
+fn no_dynamic_delete_diagnostic(span0: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn(
+        "typescript-eslint(no-dynamic-delete): Do not delete dynamically computed property keys.",
+    )
+    .with_help("Disallow using the `delete` operator on computed key expressions")
+    .with_label(span0)
+}
+
+impl Rule for NoDynamicDelete {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::UnaryExpression(expr) = node.kind() else { return };
+        if expr.operator != UnaryOperator::Delete {
+            return;
+        }
+
+        let Expression::ComputedMemberExpression(computed_expr) = &expr.argument else { return };
+
+        if computed_expr.expression.is_string_literal()
+            || computed_expr.expression.is_number_literal()
+        {
+            return;
+        }
+
+        if let Expression::UnaryExpression(unary_expr) = &computed_expr.expression {
+            if unary_expr.operator == UnaryOperator::UnaryNegation
+                && unary_expr.argument.is_number_literal()
+            {
+                return;
+            }
+        }
+        ctx.diagnostic(no_dynamic_delete_diagnostic(computed_expr.span));
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container.aaa;
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container.delete;
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container[7];
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container[-7];
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container['-Infinity'];
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container['+Infinity'];
+        	    ",
+        "
+        	const value = 1;
+        	delete value;
+        	    ",
+        "
+        	const value = 1;
+        	delete -value;
+        	    ",
+        "
+        	const container: { [i: string]: 0 } = {};
+        	delete container['aaa'];
```

Comment:
What happens if you do
```ts
// this?
delete container[('aaa')]
// or this?
delete container['aaa' + 'bbb']
```
Should we allow these cases?

Reply:
yeah you're right. `delete container[('aaa')]` should be allowed.
fixed it by using get_inner_expression function.

It seems like `delete container['aaa' + 'bbb']` should NOT be allowed. There is an invalid test case for this: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/no-dynamic-delete.test.ts#L64-L71